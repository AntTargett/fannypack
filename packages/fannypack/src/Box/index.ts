export * from './Box';
export * from './styled';
