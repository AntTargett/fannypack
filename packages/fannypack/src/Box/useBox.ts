export { useBox } from 'reakit/Box/Box';
