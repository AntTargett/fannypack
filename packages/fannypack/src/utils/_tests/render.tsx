import * as React from 'react';
import { render } from '@testing-library/react';
// @ts-ignore
import { Provider } from '../../Provider';

export default (Component: any, { theme = {} } = {}) => render(<Provider theme={theme}>{Component}</Provider>);
