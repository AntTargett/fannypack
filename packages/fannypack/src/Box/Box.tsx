import * as React from 'react';
import { Box as ReakitBox, BoxProps as ReakitBoxProps } from 'reakit';
import isPropValid from '@emotion/is-prop-valid';

import styled, { theme } from '../styled';
import { BoxThemeConfig, CSSProperties } from '../types';
import * as utils from '../utils';

export type LocalBoxProps = {
  use?: string | React.ComponentType<any>;
  children?: React.ReactNode | ((props: BoxProps) => React.ReactNode);
  /* Component-level theme overrides [Read more](TODO) */
  overrides?: BoxThemeConfig;
};
export type BoxProps = ReakitBoxProps & CSSProperties & LocalBoxProps;

export function useBoxProps(props: BoxProps | void) {
  const { use = undefined, ...restProps } = props || {};

  // Convert CSS props to a "style" prop & remove CSS props from DOM element.
  // Example input:
  // restProps = { color: 'red', backgroundColor: 'blue', isEnabled: true }
  //
  // Example output:
  // style = { color: 'red', backgroundColor: 'blue' }
  const style = utils.useStyle(restProps);

  return { use: use || StyledBox, style, ...restProps };
}

export function Box(props: BoxProps) {
  const { children, ...restProps } = props;
  const boxProps = useBoxProps(restProps);
  if (utils.isFunction(children)) {
    return <React.Fragment>{children(boxProps)}</React.Fragment>;
  }
  return (
    <StyledBox {...boxProps} as={boxProps.use !== StyledBox ? boxProps.use : undefined}>
      {children}
    </StyledBox>
  );
}

export const StyledBox = styled(ReakitBox, {
  shouldForwardProp: prop => isPropValid(prop) && !utils.isCSSProp(prop) && prop !== 'style'
})<BoxProps>`
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;

  &:focus:not(:focus-visible) {
    outline: none;
  }

  & {
    ${theme('Box.base')};
  }

  && {
    ${(props: any) => props && props.style};
  }
`;
