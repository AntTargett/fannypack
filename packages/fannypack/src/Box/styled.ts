import { Box as ReakitBox } from 'reakit';
import styled, { theme } from '../styled';
import { BoxProps } from './Box';

export const StyledBox = styled(ReakitBox)`
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;

  &:focus:not(:focus-visible) {
    outline: none;
  }

  & {
    ${theme('Box.base')};
  }
`;
