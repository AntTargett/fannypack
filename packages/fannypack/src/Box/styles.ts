import _get from 'lodash/get';
import { css, cssClass } from '../styled';
import { theme } from '../utils';

export const style = styleProps => cssClass`
  && {
    ${styleProps.style};
  }
`;

export const Box = styleProps => cssClass`
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;

  &:focus:not(:focus-visible) {
    outline: none;
  }

  ${buildVisibleAttributes(styleProps)};

  & {
    ${theme('Box.base')(styleProps)};
  }
`;

function buildVisibleAttributes(props: { hiddenBreakpoint?: string; showBreakpoint?: string }) {
  let breakpoint = props.hiddenBreakpoint || props.showBreakpoint;
  if (!breakpoint) return;

  let key: string | undefined;
  if (props.hiddenBreakpoint && breakpoint.includes('max')) {
    key = 'max-width';
  } else if (props.hiddenBreakpoint && breakpoint.includes('min')) {
    key = 'min-width';
  } else if (props.showBreakpoint && breakpoint.includes('max')) {
    key = 'min-width';
  } else if (props.showBreakpoint && breakpoint.includes('min')) {
    key = 'max-width';
  }

  let strippedBreakpoint = breakpoint;
  strippedBreakpoint = strippedBreakpoint.replace('max-', '');
  strippedBreakpoint = strippedBreakpoint.replace('min-', '');

  const minBreakpointValues: { [key: string]: number } = {
    mobile: 0,
    tablet: _get(props, 'theme.layout.mobileBreakpoint'),
    desktop: _get(props, 'theme.layout.tabletBreakpoint'),
    widescreen: _get(props, 'theme.layout.desktopBreakpoint'),
    fullHD: _get(props, 'theme.layout.widescreenBreakpoint')
  };
  let breakpointValue = _get(props, `theme.layout['${strippedBreakpoint}Breakpoint']`);
  if (props.hiddenBreakpoint && breakpoint.includes('min')) {
    breakpointValue = minBreakpointValues[strippedBreakpoint] + 1;
  }
  if (props.showBreakpoint && breakpoint.includes('min')) {
    breakpointValue = minBreakpointValues[strippedBreakpoint];
  }
  if (props.showBreakpoint && breakpoint.includes('max')) {
    breakpointValue = breakpointValue + 1;
  }

  if (!breakpoint.includes('min-') && !breakpoint.includes('max-')) {
    if (props.showBreakpoint) {
      return css`
        @media screen and (max-width: ${minBreakpointValues[breakpoint]}px) {
          display: none;
        }

        @media screen and (min-width: ${breakpointValue + 1}px) {
          display: none;
        }
      `;
    }
    return css`
      @media screen and (min-width: ${minBreakpointValues[breakpoint] + 1}px) and (max-width: ${breakpointValue}px) {
        display: none;
      }
    `;
  }
  return css`
    @media screen and (${key}: ${breakpointValue}px) {
      display: none;
    }
  `;
}
