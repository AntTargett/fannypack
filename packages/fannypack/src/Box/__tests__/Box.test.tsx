import * as React from 'react';
import { renderHook, act } from '@testing-library/react-hooks'
import { Box, useBox } from '../Box';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(<Box />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(<Box backgroundColor="red" color="white" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with a color CSS prop', () => {
    const { container } = render(<Box backgroundColor="white900" color="primary" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with a spacing CSS prop', () => {
    const { container } = render(<Box margin="major-4" paddingLeft="major-2" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with a font size CSS prop', () => {
    const { container } = render(<Box fontSize="400" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with a font weight CSS prop', () => {
    const { container } = render(<Box fontWeight="semibold" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('hook', () => {
  it('should return with box props', () => {
    const { result } = renderHook(() => useBox())
    expect(result.current).toMatchSnapshot();
  })
});
