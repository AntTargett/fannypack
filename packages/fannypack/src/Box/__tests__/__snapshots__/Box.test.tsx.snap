// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition as should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

<p
  class="emotion-0"
>
  Hello world
</p>
`;

exports[`composition hook should return with box props 1`] = `
Object {
  "style": Object {},
  "use": Object {
    "$$typeof": Symbol(react.forward_ref),
    "__emotion_base": Object {
      "$$typeof": Symbol(react.memo),
      "compare": null,
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
    },
    "__emotion_forwardProp": [Function],
    "__emotion_real": [Circular],
    "__emotion_styles": Array [
      "
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;

  &:focus:not(:focus-visible) {
    outline: none;
  }

  & {
    ",
      [Function],
      ";
  }

  && {
    ",
      [Function],
      ";
  }
",
    ],
    "defaultProps": undefined,
    "displayName": "Styled(Component)",
    "render": [Function],
    "withComponent": [Function],
  },
}
`;

exports[`composition render props should render correctly 1`] = `
<p
  use="NO_COMPONENT_SELECTOR"
>
  Hello world
</p>
`;

exports[`props should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

<div
  class="emotion-0"
/>
`;

exports[`props should render correctly with CSS props 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0.emotion-0 {
  background-color: red;
  color: white;
}

<div
  class="emotion-0"
/>
`;

exports[`props should render correctly with a color CSS prop 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0.emotion-0 {
  background-color: #d9d9d9;
  color: #444bc9;
}

<div
  class="emotion-0"
/>
`;

exports[`props should render correctly with a font size CSS prop 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0.emotion-0 {
  font-size: 1.5rem;
}

<div
  class="emotion-0"
/>
`;

exports[`props should render correctly with a font weight CSS prop 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0.emotion-0 {
  font-weight: 600;
}

<div
  class="emotion-0"
/>
`;

exports[`props should render correctly with a spacing CSS prop 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0.emotion-0 {
  margin: 2rem;
  padding-left: 1rem;
}

<div
  class="emotion-0"
/>
`;

exports[`props should render correctly with overrides 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: red;
}

<div
  class="emotion-0"
/>
`;

exports[`theming Box.base should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: red;
}

<div
  class="emotion-0"
>
  hello world
</div>
`;
